bin/dist_get

```diff
@@ -14,63 +14,65 @@ check_writeable() {
 }
 
 download() {
-	local url="$1"
-	local output="$2"
+	dl_url="$1"
+	dl_output="$2"
 
-	if [ -z "$url" ] || [ -z "$output" ]; then
-		die "download takes exactly two arguments. was given '$@'"
-	fi
+	test "$#" -eq "2" || die "download requires exactly two arguments, was given $@"
 
 	if ! check_writeable "$output"; then
 		die "download error: cannot write to $output"
 	fi
 
 	if have_binary wget; then
-		printf 'Using wget to download "%s" to "%s"\n' "$url" "$output"
+		printf '==> Using wget to download "%s" to "%s"\n' "$url" "$output"
 		wget "$url" -O "$output"
 	elif have_binary curl; then
-		printf 'Using curl to download "%s" to "%s"\n' "$url" "$output"
+		printf '==> Using curl to download "%s" to "%s"\n' "$url" "$output"
 		curl --silent "$url" > "$output"
 	elif have_binary fetch; then
-		printf 'Using fetch to download "%s" to "%s"\n' "$url" "$output"
+		printf '==> Using fetch to download "%s" to "%s"\n' "$url" "$output"
 		fetch "$url" -o "$output"
 	else
 		die "no binary found to download $url. exiting."
 	fi
+	if [ "$?" -ne 0 ]; then
+		return $?
+	fi
```

Comment:
This is wrong because when the `return $?` is executed, the value of $? is the the exit code of `[ "$?" -ne 0 ]`, which will always be true if `return $?` is executed.

You can check using for example `false ; if [ "$?" -ne 0 ]; then echo "$?" ; fi` on the command line.

By the way this is why I prefer to write tests in shell using `test` rather than using `[ ...]`. For example the bug could have been more obvious if the code had been written like this:

```
if test "$?" -ne 0; then
    return $?
fi
```

In general when you want to check "$?" after the fact it is a good idea to store "$?" into a variable as soon as possible after the command whose result you will check, like:

```
cmd
res="$?"
```

Or if you want to change the control flow depending on the result, it is often better to do it right away with `||` or `&&` just after the command, like I suggest above.

By the way `cmd || return` and `cmd || exit` do the right thing and are equivalent to `cmd || return $?` and `cmd || exit $?`.